### PR TITLE
[3006.x] Fix s3fs incorrect hash type

### LINKED ISSUE #65589: [BUG] file.managed: test mode always reports changes
**Description**

`file.managed` always reports changes in test mode, even when there are none.

**Setup**

* Salt version 3006
* Fedora server 38
* official package in Fedora with dnf
* AWS EC2 instance

**Steps to Reproduce the behavior**

It is trivially reproducible with a simple text file.

If `foo` is a text file with the contents `bar` and we for copy it to the target machine with:

```yaml
/var/tmp/foo:
  file.managed:
    - source: salt://files/foo
```

After applying, if we do `state.high state test=true`, we see:

```
----------
          ID: /var/tmp/foo
    Function: file.managed
      Result: None
     Comment: The file /var/tmp/foo is set to be changed
              Note: No changes made, actual changes may
              be different due to other states.
     Started: 21:50:33.792008
    Duration: 158.748 ms
     Changes:
              ----------
              diff:
```

**Versions Report**

<details><summary>salt --versions-report</summary>

```yaml
Salt Version:
          Salt: 3006.4

Python Version:
        Python: 3.11.6 (main, Oct  3 2023, 00:00:00) [GCC 13.2.1 20230728
(Red Hat 13.2.1-1)]

Dependency Versions:
          cffi: 1.15.1
      cherrypy: unknown
      dateutil: 2.8.2
     docker-py: Not Installed
         gitdb: Not Installed
     gitpython: Not Installed
        Jinja2: 3.0.3
       libgit2: Not Installed
  looseversion: 1.3.0
      M2Crypto: Not Installed
          Mako: Not Installed
       msgpack: 1.0.4
  msgpack-pure: Not Installed
  mysql-python: Not Installed
     packaging: 23.0
     pycparser: 2.20
      pycrypto: Not Installed
  pycryptodome: 3.19.0
        pygit2: Not Installed
  python-gnupg: Not Installed
        PyYAML: 6.0
         PyZMQ: 24.0.1
smmap: Not Installed
       timelib: Not Installed
       Tornado: 4.5.3
           ZMQ: 4.3.4

System Versions:
          dist: fedora 38
        locale: utf-8
       machine: x86_64
       release: 6.5.8-200.fc38.x86_64
        system: Linux
       version: Fedora Linux 38
```
</details>

### PR DESCRIPTION
**Pull request by dead10ck**
### What does this PR do?

Fixes #65589

### Merge requirements satisfied?

- [ ] Docs
- [x] Changelog - https://docs.saltproject.io/en/master/topics/development/changelog.html
- [x] Tests written/updated

### CODE REVIEW

**Comment by dead10ck** [comment]
cc @whytewolf - thanks again for helping debug

**Comment by dead10ck** [comment]
Sorry this may take a lot of iterations because the precommit checks don't work on my local system because they're incompatible with Python 3.11

**Comment by dead10ck** [comment]
Do the remaining errors have anything to do with my change?

**Comment by whytewolf** [comment]
> Do the remaining errors have anything to do with my change

it doesn't look like it. the tests didn't even run. i know @s0undt3ch is working on some fixes for the test suite. so might have to wait until those are fixed and we can rebase after that.

**Comment by whytewolf** [comment]
> Is there an issue I can subscribe to, just so I know when this PR can potentially be unblocked?

I just noticed this went in against master. since this is a bugfix it should be against 3006.x sorry I didn't notice that earlier. 

unfortunately there is not a issue setup to track the things @s0undt3ch is fixing. it is just a lot of back-end work mostly. and GitHub actions fixes.

**Comment by dead10ck** [comment]
Ah ok, no problem, I'll rebase

File: noxfile.py
Comment (s0undt3ch) [inline noxfile.py:1]: You should either get a newer `nox` installed locally, or submit these changes separately.

**Review by s0undt3ch — CHANGES_REQUESTED** [review]

File: tests/pytests/unit/fileserver/test_s3fs.py
Comment (dead10ck) [reply tests/pytests/unit/fileserver/test_s3fs.py:67]: Very confused by this comment. I was just mutating the variable, but a lint failed and explicitly told me to use `@patch` instead. There are hundreds of preexisting uses of `@patch`. The docs on writing tests say to use it. Even the article you linked to says to use it:

> we often were using a mock in these cases, so we replaced them all with mock.patch(…). Lesson: Use mock.patch… it’s designed for this sort of thing and makes cleaning up easier.

File: noxfile.py
Comment (dead10ck) [reply noxfile.py:1]: This was changed because even in the newest version of nox as of this writing, an exception is thrown if you pass `--no-install`. Turns out that despite its name, `run_always` doesn't run if you pass this flag, so it just returns `None`.

I can remove it from this PR if you'd rather not have it though.

File: tests/pytests/unit/fileserver/test_s3fs.py
Comment (dead10ck) [reply tests/pytests/unit/fileserver/test_s3fs.py:8]: Yes really. I'm not totally positive, but their [docs](https://docs.getmoto.org/en/latest/docs/getting_started.html#how-do-i-avoid-tests-from-mutating-my-real-infrastructure) state it pretty clearly. I'm guessing it's because there are either some cases or versions or something that run authentication initialization during the import, so importing moto first ensures we can always beat it to the punch, so to speak.

I generally found moto very finicky to work with during the implementation of these tests. Much care is needed to avoid nesting scopes and such. I guess that's just mocks though.

File: tests/pytests/unit/fileserver/test_s3fs.py
Comment (dead10ck) [reply tests/pytests/unit/fileserver/test_s3fs.py:68]: It's to avoid repeating the same `@patch` for every test. They all need it.

Comment (s0undt3ch) [reply tests/pytests/unit/fileserver/test_s3fs.py:68]: Just use a module scoped fixture

Comment (dead10ck) [reply tests/pytests/unit/fileserver/test_s3fs.py:68]: I'm not really sure what you mean. It's not a fixture. We're modifying a global variable in the Python module.

Comment (s0undt3ch) [reply tests/pytests/unit/fileserver/test_s3fs.py:68]:
```python
@pytest.fixture(scope="module", autouse=True)
def _patch_cache_expire():
    with patch("salt.fileserver.s3fs.S3_CACHE_EXPIRE", -1):
        yield
```

File: tests/pytests/unit/fileserver/test_s3fs.py
Comment (dead10ck) [reply tests/pytests/unit/fileserver/test_s3fs.py:20]: I tried this initially but couldn't get it to work. I'm not sure if there's some weird interaction with pytest and moto, but when it's a fixture, moto doesn't seem to mock properly and we end up getting errors like:

```
E           salt.exceptions.CommandExecutionError: Failed s3 operation. http-400: b'The method is not implemented'
```

File: tests/pytests/unit/fileserver/test_s3fs.py
Comment (lkubb) [reply tests/pytests/unit/fileserver/test_s3fs.py]: You can do the same as with the class fixture below. The problem is using the patch decorator in the module's global scope, it's fine inside a fixture/test function.

See the above article at

> Create no mocks before your test starts. Some of our tests defined top-level mocks in tests using the @patch decorator from the mock library as follows

File: tests/pytests/unit/fileserver/test_s3fs.py
Comment (dead10ck) [reply tests/pytests/unit/fileserver/test_s3fs.py]: Never mind, I figured out what the problem was. I changed these to fixtures.

File: salt/fileserver/s3fs.py
Comment (s0undt3ch) [inline salt/fileserver/s3fs.py:130]: Thank You!

File: tests/pytests/unit/fileserver/test_s3fs.py
Comment (s0undt3ch) [inline tests/pytests/unit/fileserver/test_s3fs.py:30]:
```suggestion
@pytest.fixture
```
This was fine as it was, because `function` scope is the default.
Not going to block the PR because of this tough.

File: tests/pytests/unit/fileserver/test_s3fs.py
Comment (s0undt3ch) [inline tests/pytests/unit/fileserver/test_s3fs.py:51]:
```suggestion
@pytest.fixture(scope="function")
def s3(bucket, aws_creds):
    # these can't be fixtures, as moto docs suggest, since they would introduce a
    # separate scope of mock than the s3fs module's scope
    def _s3():
        conn = boto3.client("s3", **aws_creds)
        conn.create_bucket(Bucket=bucket)
        return conn

    return _s3
```
Before the comment made sense where it was, now that `s3` is actually a fixture, better move the comment inside it(or remove the comment?)

File: tests/pytests/unit/fileserver/test_s3fs.py
Comment (s0undt3ch) [inline tests/pytests/unit/fileserver/test_s3fs.py:146]:
```suggestion
@pytest.fixture
def _mock_s3():
    with mock_s3():
        yield


@pytest.mark.usefixture("_mock_s3")
def test_update(bucket, s3):
    """Tests that files get downloaded from s3 to the local cache."""

    keys = {
        "top.sls": {"content": yaml.dump({"base": {"*": ["foo"]}})},
        "foo.sls": {"content": yaml.dump({"nginx": {"pkg.installed": []}})},
        "files/nginx.conf": {"content": "server {}"},
    }

    conn = s3()
    make_keys(bucket, conn, keys)
    s3fs.update()
    verify_cache(bucket, keys)

    # make a modification and update again - verify the change is retrieved
    keys["top.sls"]["content"] = yaml.dump({"base": {"*": ["foo", "bar"]}})
    make_keys(bucket, conn, keys)
    s3fs.update()
    verify_cache(bucket, keys)

    # TODO: fix

    # delete_file = "files/nginx.conf"
    # del keys[delete_file]
    # conn.delete_object(Bucket=bucket, Key=delete_file)

    # s3fs.update()
    # verify_cache(bucket, keys)

    # cache_file = s3fs._get_cached_file_name(bucket, "base", delete_file)
    # assert not os.path.exists(cache_file)


@pytest.mark.usefixture("_mock_s3")
def test_s3_hash(bucket, s3):
    """Verifies that s3fs hashes files correctly."""

    keys = {
        "top.sls": {"content": yaml.dump({"base": {"*": ["foo"]}})},
        "foo.sls": {"content": yaml.dump({"nginx": {"pkg.installed": []}})},
        "files/nginx.conf": {"content": "server {}"},
    }

    conn = s3()
    make_keys(bucket, conn, keys)
    s3fs.update()

    for key, item in keys.items():
        cached_file_path = s3fs._get_cached_file_name(bucket, "base", key)
        item["hash"] = salt.utils.hashutils.get_hash(
            cached_file_path, s3fs.S3_HASH_TYPE
        )
        item["cached_file_path"] = cached_file_path

    load = {"saltenv": "base"}
    fnd = {"bucket": bucket}

    for key, item in keys.items():
        fnd["path"] = item["cached_file_path"]
        actual_hash = s3fs.file_hash(load, fnd)
        assert s3fs.S3_HASH_TYPE == actual_hash["hash_type"]
        assert item["hash"] == actual_hash["hsum"]
```

http://docs.getmoto.org/en/latest/docs/getting_started.html?highlight=mock_s3#context-manager

Or, alternatively(to not have to use `@pytest.mark.usefixture("_mock_s3")`:
```python
@pytest.fixture
def bucket():
    with mock_s3():
        yield "bucket-name"
```

**Review by s0undt3ch — CHANGES_REQUESTED** [review]

File: tests/pytests/unit/fileserver/test_s3fs.py
Comment (dead10ck) [reply tests/pytests/unit/fileserver/test_s3fs.py:30]: Yeah, it's not a totally necessary change; I just prefer to be explicit, rather than rely on defaults. This PR is a good example of why 🙂

Up to you though, I don't mind changing it back if you'd like

File: tests/pytests/unit/fileserver/test_s3fs.py
Comment (dead10ck) [reply tests/pytests/unit/fileserver/test_s3fs.py:51]: Ah yep, forgot to update the comment. I'll just remove it

File: tests/pytests/unit/fileserver/test_s3fs.py
Comment (s0undt3ch) [reply tests/pytests/unit/fileserver/test_s3fs.py:30]: All good

File: tests/pytests/unit/fileserver/test_s3fs.py
Comment (dead10ck) [reply tests/pytests/unit/fileserver/test_s3fs.py]: Done

**Comment by dead10ck** [comment]
Ok I think I got all the comments

File: tests/pytests/unit/fileserver/test_s3fs.py
Comment (lkubb) [inline tests/pytests/unit/fileserver/test_s3fs.py:50]: question: Why does this fixture return a factory instead of the object? Doing the setup in the fixture function itself could DRY up the tests and is how `pytest` envisions fixtures afaik.

```suggestion
@pytest.fixture
def files_present():
    return {
        "top.sls": {"content": yaml.dump({"base": {"*": ["foo"]}})},
        "foo.sls": {"content": yaml.dump({"nginx": {"pkg.installed": []}})},
        "files/nginx.conf": {"content": "server {}"},
    }


@pytest.fixture(scope="function")
def s3(bucket, aws_creds, files_present):
    conn = boto3.client("s3", **aws_creds)
    conn.create_bucket(Bucket=bucket)
    make_keys(bucket, conn, files_present)
    s3fs.update()
    verify_cache(bucket, files_present)
    return conn
```

File: tests/pytests/unit/fileserver/test_s3fs.py
Comment (lkubb) [inline tests/pytests/unit/fileserver/test_s3fs.py:101]: suggestion: With the above, this can be reduced to

```suggestion
def test_update(bucket, s3, files_present):
    """Tests that files get downloaded from s3 to the local cache."""
    # make a modification and update again - verify the change is retrieved
    files_present["top.sls"]["content"] = yaml.dump({"base": {"*": ["foo", "bar"]}})
    make_keys(bucket, s3, files_present)
    s3fs.update()
    verify_cache(bucket, files_present)
```

File: tests/pytests/unit/fileserver/test_s3fs.py
Comment (lkubb) [inline tests/pytests/unit/fileserver/test_s3fs.py:139]: ... and this to

```suggestion
def test_s3_hash(bucket, s3, files_present):
    """Verifies that s3fs hashes files correctly."""

    for key, item in files_present.items():
        cached_file_path = s3fs._get_cached_file_name(bucket, "base", key)
        item["hash"] = salt.utils.hashutils.get_hash(
            cached_file_path, s3fs.S3_HASH_TYPE
        )
        item["cached_file_path"] = cached_file_path

    load = {"saltenv": "base"}
    fnd = {"bucket": bucket}

    for key, item in files_present.items():
```

**Review by lkubb — APPROVED** [review]
The `_patch_cache_expire` fixture is correct imho.

Just a suggestion for the general fixture setup, not sure if there is a reason it's done like this. Note that I have no power here, just tried to help before. :)

(I couldn't run the tests due to an exception by `moto`)

File: tests/pytests/unit/fileserver/test_s3fs.py
Comment (dead10ck) [reply tests/pytests/unit/fileserver/test_s3fs.py]: Actually you're right! It was necessary before I got the bucket being created in a fixture as well, but it's not any more. I changed it to just a function scoped fixture instead of a factory.

However, I think making the particular test files that get uploaded into a fixture takes it a bit too far, since they are particular test inputs. They happen to be the same here, but even in #65614, which will follow this one, the files change slightly to test another case.

Comment (lkubb) [reply tests/pytests/unit/fileserver/test_s3fs.py]: You could use indirect parametrization in that case, but admittedly that gets increasingly complex...

```python
@pytest.fixture(params=("default_files",))
def files_present(request):
    return request.getfixturevalue(request.param)


@pytest.fixture
def default_files():
    return {
        "top.sls": {"content": yaml.dump({"base": {"*": ["foo"]}})},
        "foo.sls": {"content": yaml.dump({"nginx": {"pkg.installed": []}})},
        "files/nginx.conf": {"content": "server {}"},
    }


@pytest.fixture
def other_files():
    return {
        "top.sls": {"content": yaml.dump({"base": {"*": ["bar"]}})},
        "bar.sls": {"content": yaml.dump({"nginx": {"pkg.removed": []}})},
    }


@pytest.mark.parametrize("files_present", ("other_files",), indirect=True)
def test_something_else(s3, files_present):
    """ ... """
```

Comment (dead10ck) [reply tests/pytests/unit/fileserver/test_s3fs.py]: Yeah this isn't really helping readability. If it were the same files repeated half a dozen times it would be a different story, but the tradeoff isn't worth it yet, in my opinion.

File: tests/pytests/unit/fileserver/test_s3fs.py
Comment (s0undt3ch) [inline tests/pytests/unit/fileserver/test_s3fs.py:40]:
```suggestion
def configure_loader_modules(tmp_path, bucket):
    opts = {
        "cachedir": tmp_path,
        "s3.buckets": {"base": [bucket]},
        "s3.location": "us-east-1",
    }
    utils = {"s3.query": salt.utils.s3.query}

    # [HACK] to always invalidate the cache and re-download on every update
    with patch("salt.fileserver.s3fs.S3_CACHE_EXPIRE", -1):
        yield {s3fs: {"__opts__": opts, "__utils__": utils}}
```

File: tests/pytests/unit/fileserver/test_s3fs.py
Comment (s0undt3ch) [inline tests/pytests/unit/fileserver/test_s3fs.py:79]:
```suggestion
```
See above for this hack still getting used

**Review by s0undt3ch — CHANGES_REQUESTED** [review]

File: tests/pytests/unit/fileserver/test_s3fs.py
Comment (dead10ck) [reply tests/pytests/unit/fileserver/test_s3fs.py]: Done

**Review by s0undt3ch — DISMISSED** [review]

**Review by twangboy — APPROVED** [review]

**Comment by dead10ck** [comment]
Is there anything else this PR needs?

**Review by twangboy — DISMISSED** [review]

**Comment by twangboy** [comment]
Just rebased... lets see if that fixes the failing tests.

**Comment by dead10ck** [comment]
Ok I'll do that. Should I change the base branch to 3007.x now?

**Comment by whytewolf** [comment]
> Ok I'll do that. Should I change the base branch to 3007.x now?

3006.x is an lts, so bug fixes should still be going into 3006.x. so no please do not move to 3007. We will handle that separately once merged into 3006.x

Though can you resolve that file conflict? not sure what happened there.

**Comment by dead10ck** [comment]
Oh sorry, I saw your comment too late. I'll move it back

**Comment by dead10ck** [comment]
Ok it would be all set now. The conflict was #65625

**Review by twangboy — DISMISSED** [review]

**Comment by dead10ck** [comment]
> The test failures look valid

Which failures are you looking at? This looks like the test environment is misconfigured.

```
           =========================== short test summary info ============================
           SKIPPED [1] tests/pytests/unit/crypt/test_crypt_m2crypto.py:10: could not import 'M2Crypto': No module named 'M2Crypto'
           FAILED tests/pytests/unit/fileserver/test_s3fs.py::test_update - ValueError: [digital envelope routines] unsupported
           FAILED tests/pytests/unit/fileserver/test_s3fs.py::test_s3_hash - ValueError: [digital envelope routines] unsupported
```

**Comment by dead10ck** [comment]
Ok I think I added the right marker to skip the tests on fips systems

**Review by twangboy — APPROVED** [review]

**Comment by dead10ck** [comment]
The tests that are failing now look unrelated. Missing modules in tests for redis on MacOS.

**Review by s0undt3ch — APPROVED** [review]

**Pull request merged**
